File: app/api/v1/endpoints/financial.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import func, and_, or_
from pydantic import BaseModel
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal
from app.core.database import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.models.financial import FinancialRecord, TransactionType, PaymentStatus
from app.models.client import Client
from app.models.process import Process

router = APIRouter()


class FinancialRecordCreate(BaseModel):
    transaction_type: str  # receivable, received, expense
    description: str
    amount: float
    payment_status: Optional[str] = "pending"
    due_date: Optional[date] = None
    payment_date: Optional[date] = None
    payment_method: Optional[str] = None
    invoice_number: Optional[str] = None
    notes: Optional[str] = None
    client_id: Optional[int] = None
    process_id: Optional[int] = None
    category: Optional[str] = None


class FinancialRecordUpdate(BaseModel):
    transaction_type: Optional[str] = None
    description: Optional[str] = None
    amount: Optional[float] = None
    payment_status: Optional[str] = None
    due_date: Optional[date] = None
    payment_date: Optional[date] = None
    payment_method: Optional[str] = None
    invoice_number: Optional[str] = None
    notes: Optional[str] = None
    client_id: Optional[int] = None
    process_id: Optional[int] = None
    category: Optional[str] = None


@router.post("/")
async def create_financial_record(
    record_data: FinancialRecordCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Cadastrar registro financeiro
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    record = FinancialRecord(
        transaction_type=TransactionType(record_data.transaction_type),
        description=record_data.description,
        amount=Decimal(str(record_data.amount)),
        payment_status=PaymentStatus(record_data.payment_status),
        due_date=record_data.due_date,
        payment_date=record_data.payment_date,
        payment_method=record_data.payment_method,
        invoice_number=record_data.invoice_number,
        notes=record_data.notes,
        client_id=record_data.client_id,
        process_id=record_data.process_id,
        law_firm_id=law_firm_id
    )

    db.add(record)
    db.commit()
    db.refresh(record)

    return {
        "id": record.id,
        "transaction_type": record.transaction_type.value,
        "description": record.description,
        "amount": float(record.amount),
        "payment_status": record.payment_status.value,
        "due_date": record.due_date.isoformat() if record.due_date else None,
        "payment_date": record.payment_date.isoformat() if record.payment_date else None,
        "created_at": record.created_at.isoformat() if record.created_at else None
    }


@router.get("/")
async def list_financial_records(
    transaction_type: Optional[str] = Query(None),
    payment_status: Optional[str] = Query(None),
    client_id: Optional[int] = Query(None),
    process_id: Optional[int] = Query(None),
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Listar registros financeiros com filtros
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    query = db.query(FinancialRecord).filter(FinancialRecord.law_firm_id == law_firm_id)

    # Apply filters
    if transaction_type:
        query = query.filter(FinancialRecord.transaction_type == transaction_type)
    if payment_status:
        query = query.filter(FinancialRecord.payment_status == payment_status)
    if client_id:
        query = query.filter(FinancialRecord.client_id == client_id)
    if process_id:
        query = query.filter(FinancialRecord.process_id == process_id)
    if start_date:
        query = query.filter(FinancialRecord.due_date >= start_date)
    if end_date:
        query = query.filter(FinancialRecord.due_date <= end_date)

    records = query.order_by(FinancialRecord.due_date.desc()).all()

    result = []
    for r in records:
        client_name = None
        if r.client_id:
            client = db.query(Client).filter(Client.id == r.client_id).first()
            if client:
                client_name = client.name

        process_number = None
        if r.process_id:
            process = db.query(Process).filter(Process.id == r.process_id).first()
            if process:
                process_number = process.process_number

        result.append({
            "id": r.id,
            "transaction_type": r.transaction_type.value,
            "description": r.description,
            "amount": float(r.amount),
            "payment_status": r.payment_status.value,
            "due_date": r.due_date.isoformat() if r.due_date else None,
            "payment_date": r.payment_date.isoformat() if r.payment_date else None,
            "payment_method": r.payment_method,
            "invoice_number": r.invoice_number,
            "notes": r.notes,
            "client_id": r.client_id,
            "client_name": client_name,
            "process_id": r.process_id,
            "process_number": process_number,
            "created_at": r.created_at.isoformat() if r.created_at else None
        })

    return result


@router.get("/summary")
async def get_financial_summary(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Obter resumo financeiro
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    # Total a receber (pendente)
    total_receivable = db.query(func.sum(FinancialRecord.amount)).filter(
        FinancialRecord.law_firm_id == law_firm_id,
        FinancialRecord.transaction_type == TransactionType.RECEIVABLE,
        FinancialRecord.payment_status == PaymentStatus.PENDING
    ).scalar() or 0

    # Total recebido
    total_received = db.query(func.sum(FinancialRecord.amount)).filter(
        FinancialRecord.law_firm_id == law_firm_id,
        or_(
            FinancialRecord.transaction_type == TransactionType.RECEIVED,
            and_(
                FinancialRecord.transaction_type == TransactionType.RECEIVABLE,
                FinancialRecord.payment_status == PaymentStatus.PAID
            )
        )
    ).scalar() or 0

    # Total despesas
    total_expenses = db.query(func.sum(FinancialRecord.amount)).filter(
        FinancialRecord.law_firm_id == law_firm_id,
        FinancialRecord.transaction_type == TransactionType.EXPENSE,
        FinancialRecord.payment_status == PaymentStatus.PAID
    ).scalar() or 0

    # Total atrasado
    total_overdue = db.query(func.sum(FinancialRecord.amount)).filter(
        FinancialRecord.law_firm_id == law_firm_id,
        FinancialRecord.payment_status == PaymentStatus.OVERDUE
    ).scalar() or 0

    # Despesas pendentes
    total_expenses_pending = db.query(func.sum(FinancialRecord.amount)).filter(
        FinancialRecord.law_firm_id == law_firm_id,
        FinancialRecord.transaction_type == TransactionType.EXPENSE,
        FinancialRecord.payment_status == PaymentStatus.PENDING
    ).scalar() or 0

    return {
        "total_receivable": float(total_receivable),
        "total_received": float(total_received),
        "total_expenses": float(total_expenses),
        "total_overdue": float(total_overdue),
        "total_expenses_pending": float(total_expenses_pending),
        "balance": float(total_received) - float(total_expenses)
    }


@router.get("/{record_id}")
async def get_financial_record(
    record_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Obter detalhes do registro financeiro
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    record = db.query(FinancialRecord).filter(
        FinancialRecord.id == record_id,
        FinancialRecord.law_firm_id == law_firm_id
    ).first()

    if not record:
        raise HTTPException(status_code=404, detail="Registro financeiro não encontrado")

    client_name = None
    if record.client_id:
        client = db.query(Client).filter(Client.id == record.client_id).first()
        if client:
            client_name = client.name

    process_number = None
    if record.process_id:
        process = db.query(Process).filter(Process.id == record.process_id).first()
        if process:
            process_number = process.process_number

    return {
        "id": record.id,
        "transaction_type": record.transaction_type.value,
        "description": record.description,
        "amount": float(record.amount),
        "payment_status": record.payment_status.value,
        "due_date": record.due_date.isoformat() if record.due_date else None,
        "payment_date": record.payment_date.isoformat() if record.payment_date else None,
        "payment_method": record.payment_method,
        "invoice_number": record.invoice_number,
        "notes": record.notes,
        "client_id": record.client_id,
        "client_name": client_name,
        "process_id": record.process_id,
        "process_number": process_number,
        "created_at": record.created_at.isoformat() if record.created_at else None
    }


@router.put("/{record_id}")
async def update_financial_record(
    record_id: int,
    record_data: FinancialRecordUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Atualizar registro financeiro
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    record = db.query(FinancialRecord).filter(
        FinancialRecord.id == record_id,
        FinancialRecord.law_firm_id == law_firm_id
    ).first()

    if not record:
        raise HTTPException(status_code=404, detail="Registro financeiro não encontrado")

    # Update fields if provided
    if record_data.transaction_type:
        record.transaction_type = TransactionType(record_data.transaction_type)
    if record_data.description:
        record.description = record_data.description
    if record_data.amount is not None:
        record.amount = Decimal(str(record_data.amount))
    if record_data.payment_status:
        record.payment_status = PaymentStatus(record_data.payment_status)
    if record_data.due_date is not None:
        record.due_date = record_data.due_date
    if record_data.payment_date is not None:
        record.payment_date = record_data.payment_date
    if record_data.payment_method:
        record.payment_method = record_data.payment_method
    if record_data.invoice_number:
        record.invoice_number = record_data.invoice_number
    if record_data.notes:
        record.notes = record_data.notes
    if record_data.client_id is not None:
        record.client_id = record_data.client_id
    if record_data.process_id is not None:
        record.process_id = record_data.process_id

    db.commit()
    db.refresh(record)

    return {"message": "Registro financeiro atualizado com sucesso", "id": record.id}


@router.delete("/{record_id}")
async def delete_financial_record(
    record_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Deletar registro financeiro
    """
    law_firm_id = current_user.law_firm_id if current_user.law_firm_id else current_user.id

    record = db.query(FinancialRecord).filter(
        FinancialRecord.id == record_id,
        FinancialRecord.law_firm_id == law_firm_id
    ).first()

    if not record:
        raise HTTPException(status_code=404, detail="Registro financeiro não encontrado")

    db.delete(record)
    db.commit()

    return {"message": "Registro financeiro deletado com sucesso"}

File: app/static/js/financial.js
// Financial Module
let currentFinancialRecord = null;

// Format currency
function formatCurrency(value) {
    return new Intl.NumberFormat('pt-BR', {
        style: 'currency',
        currency: 'BRL'
    }).format(value);
}

// Format date
function formatDate(dateString) {
    if (!dateString) return '-';
    const date = new Date(dateString);
    return date.toLocaleDateString('pt-BR');
}

// Load Financial Summary
async function loadFinancialSummary() {
    try {
        const response = await fetch(`${API_URL}/api/v1/financial/summary`, {
            headers: getAuthHeaders()
        });

        if (response.ok) {
            const summary = await response.json();
            document.getElementById('stats-total-received').textContent = formatCurrency(summary.total_received || 0);
            document.getElementById('stats-total-receivable').textContent = formatCurrency(summary.total_receivable || 0);
            document.getElementById('stats-total-expenses').textContent = formatCurrency(summary.total_expenses || 0);
            document.getElementById('stats-total-overdue').textContent = formatCurrency(summary.total_overdue || 0);
            document.getElementById('stats-expenses-pending').textContent = formatCurrency(summary.total_expenses_pending || 0);
            document.getElementById('stats-balance').textContent = formatCurrency(summary.balance || 0);
        }
    } catch (error) {
        console.error('Error loading financial summary:', error);
    }
}

// Load Financial Records
async function loadFinancialRecords(filters = {}) {
    const container = document.getElementById('financial-list');
    container.innerHTML = '<p class="text-gray-500 text-center py-8">Carregando...</p>';

    try {
        // Build query string
        const params = new URLSearchParams();
        if (filters.transaction_type) params.append('transaction_type', filters.transaction_type);
        if (filters.payment_status) params.append('payment_status', filters.payment_status);
        if (filters.start_date) params.append('start_date', filters.start_date);
        if (filters.end_date) params.append('end_date', filters.end_date);

        const queryString = params.toString();
        const url = `${API_URL}/api/v1/financial/${queryString ? '?' + queryString : ''}`;

        const response = await fetch(url, {
            headers: getAuthHeaders()
        });

        if (response.ok) {
            const records = await response.json();

            if (records.length === 0) {
                container.innerHTML = '<p class="text-gray-500 text-center py-8">Nenhum registro financeiro encontrado</p>';
                return;
            }

            container.innerHTML = records.map(record => {
                const typeLabels = {
                    'receivable': 'A Receber',
                    'received': 'Recebido',
                    'expense': 'Despesa'
                };

                const statusLabels = {
                    'pending': 'Pendente',
                    'paid': 'Pago',
                    'overdue': 'Atrasado',
                    'cancelled': 'Cancelado'
                };

                const typeColors = {
                    'receivable': 'text-yellow-600',
                    'received': 'text-green-600',
                    'expense': 'text-red-600'
                };

                const statusColors = {
                    'pending': 'bg-yellow-100 text-yellow-800',
                    'paid': 'bg-green-100 text-green-800',
                    'overdue': 'bg-red-100 text-red-800',
                    'cancelled': 'bg-gray-100 text-gray-800'
                };

                return `
                    <div class="border border-gray-200 rounded-lg p-4 hover:shadow-md transition">
                        <div class="flex justify-between items-start">
                            <div class="flex-1">
                                <div class="flex items-center mb-2">
                                    <h4 class="text-lg font-semibold text-gray-800 mr-3">${record.description}</h4>
                                    <span class="inline-block px-2 py-1 text-xs font-semibold rounded ${statusColors[record.payment_status]}">
                                        ${statusLabels[record.payment_status]}
                                    </span>
                                </div>
                                <p class="text-2xl font-bold ${typeColors[record.transaction_type]} mb-2">
                                    ${formatCurrency(record.amount)}
                                </p>
                                <div class="grid grid-cols-2 gap-2 text-sm text-gray-600">
                                    <div>
                                        <span class="font-medium">Tipo:</span> ${typeLabels[record.transaction_type]}
                                    </div>
                                    ${record.due_date ? `
                                        <div>
                                            <span class="font-medium">Vencimento:</span> ${formatDate(record.due_date)}
                                        </div>
                                    ` : ''}
                                    ${record.payment_date ? `
                                        <div>
                                            <span class="font-medium">Pagamento:</span> ${formatDate(record.payment_date)}
                                        </div>
                                    ` : ''}
                                    ${record.payment_method ? `
                                        <div>
                                            <span class="font-medium">Método:</span> ${record.payment_method}
                                        </div>
                                    ` : ''}
                                    ${record.client_name ? `
                                        <div>
                                            <span class="font-medium">Cliente:</span> ${record.client_name}
                                        </div>
                                    ` : ''}
                                    ${record.process_number ? `
                                        <div>
                                            <span class="font-medium">Processo:</span> ${record.process_number}
                                        </div>
                                    ` : ''}
                                    ${record.invoice_number ? `
                                        <div>
                                            <span class="font-medium">NF:</span> ${record.invoice_number}
                                        </div>
                                    ` : ''}
                                </div>
                                ${record.notes ? `
                                    <p class="text-sm text-gray-600 mt-2 italic">${record.notes}</p>
                                ` : ''}
                            </div>
                            <div class="flex flex-col space-y-2 ml-4">
                                <button onclick="editFinancialRecord(${record.id})" class="text-blue-600 hover:text-blue-800" title="Editar">
                                    <i class="fas fa-edit"></i>
                                </button>
                                <button onclick="deleteFinancialRecord(${record.id})" class="text-red-600 hover:text-red-800" title="Excluir">
                                    <i class="fas fa-trash"></i>
                                </button>
                            </div>
                        </div>
                    </div>
                `;
            }).join('');
        } else {
            container.innerHTML = '<p class="text-red-500 text-center py-8">Erro ao carregar registros financeiros</p>';
        }
    } catch (error) {
        console.error('Error loading financial records:', error);
        container.innerHTML = '<p class="text-red-500 text-center py-8">Erro de conexão</p>';
    }
}

// Open Financial Modal
async function openFinancialModal(type = 'receivable', recordId = null) {
    currentFinancialRecord = recordId;

    const modal = document.getElementById('financialModal');
    const form = document.getElementById('financialForm');
    const title = document.getElementById('financialModalTitle');
    const submitBtn = document.getElementById('financialSubmitBtn');

    // Reset form
    form.reset();
    document.getElementById('financialRecordId').value = '';

    // Load clients and processes for selects
    await loadFinancialSelects();

    if (recordId) {
        // Edit mode
        title.textContent = 'Editar Registro Financeiro';
        submitBtn.textContent = 'Atualizar';

        try {
            const response = await fetch(`${API_URL}/api/v1/financial/${recordId}`, {
                headers: getAuthHeaders()
            });

            if (response.ok) {
                const record = await response.json();

                // Fill form
                document.getElementById('financialRecordId').value = record.id;
                form.querySelector('[name="transaction_type"]').value = record.transaction_type;
                form.querySelector('[name="description"]').value = record.description;
                form.querySelector('[name="amount"]').value = record.amount;
                form.querySelector('[name="payment_status"]').value = record.payment_status;
                form.querySelector('[name="due_date"]').value = record.due_date || '';
                form.querySelector('[name="payment_date"]').value = record.payment_date || '';
                form.querySelector('[name="payment_method"]').value = record.payment_method || '';
                form.querySelector('[name="invoice_number"]').value = record.invoice_number || '';
                form.querySelector('[name="client_id"]').value = record.client_id || '';
                form.querySelector('[name="process_id"]').value = record.process_id || '';
                form.querySelector('[name="notes"]').value = record.notes || '';
            }
        } catch (error) {
            console.error('Error loading financial record:', error);
            alert('Erro ao carregar registro financeiro');
            return;
        }
    } else {
        // Create mode
        title.textContent = type === 'expense' ? 'Nova Despesa' : 'Nova Receita';
        submitBtn.textContent = 'Criar';
        form.querySelector('[name="transaction_type"]').value = type;
    }

    modal.classList.remove('hidden');
}

// Close Financial Modal
function closeFinancialModal() {
    document.getElementById('financialModal').classList.add('hidden');
    currentFinancialRecord = null;
}

// Load clients and processes for selects
async function loadFinancialSelects() {
    try {
        // Load clients
        const clientsRes = await fetch(`${API_URL}/api/v1/clients/`, {
            headers: getAuthHeaders()
        });
        if (clientsRes.ok) {
            const clients = await clientsRes.json();
            const clientSelect = document.getElementById('financialClientSelect');
            clientSelect.innerHTML = '<option value="">Nenhum</option>' +
                clients.map(c => `<option value="${c.id}">${c.name}</option>`).join('');
        }

        // Load processes
        const processesRes = await fetch(`${API_URL}/api/v1/processes/`, {
            headers: getAuthHeaders()
        });
        if (processesRes.ok) {
            const processes = await processesRes.json();
            const processSelect = document.getElementById('financialProcessSelect');
            processSelect.innerHTML = '<option value="">Nenhum</option>' +
                processes.map(p => `<option value="${p.id}">${p.process_number} - ${p.subject}</option>`).join('');
        }
    } catch (error) {
        console.error('Error loading selects:', error);
    }
}

// Handle form submission
document.getElementById('financialForm').addEventListener('submit', async (e) => {
    e.preventDefault();

    const formData = new FormData(e.target);
    const data = {
        transaction_type: formData.get('transaction_type'),
        description: formData.get('description'),
        amount: parseFloat(formData.get('amount')),
        payment_status: formData.get('payment_status'),
        due_date: formData.get('due_date') || null,
        payment_date: formData.get('payment_date') || null,
        payment_method: formData.get('payment_method') || null,
        invoice_number: formData.get('invoice_number') || null,
        client_id: formData.get('client_id') ? parseInt(formData.get('client_id')) : null,
        process_id: formData.get('process_id') ? parseInt(formData.get('process_id')) : null,
        notes: formData.get('notes') || null
    };

    const recordId = document.getElementById('financialRecordId').value;

    try {
        const url = recordId
            ? `${API_URL}/api/v1/financial/${recordId}`
            : `${API_URL}/api/v1/financial/`;

        const method = recordId ? 'PUT' : 'POST';

        const response = await fetch(url, {
            method: method,
            headers: getAuthHeaders(),
            body: JSON.stringify(data)
        });

        if (response.ok) {
            alert(recordId ? 'Registro atualizado com sucesso!' : 'Registro criado com sucesso!');
            closeFinancialModal();
            loadFinancialSummary();
            loadFinancialRecords();
        } else {
            const error = await response.json();
            alert('Erro: ' + (error.detail || 'Erro desconhecido'));
        }
    } catch (error) {
        console.error('Error saving financial record:', error);
        alert('Erro de conexão');
    }
});

// Edit Financial Record
function editFinancialRecord(id) {
    openFinancialModal('receivable', id);
}

// Delete Financial Record
async function deleteFinancialRecord(id) {
    if (confirm('Tem certeza que deseja excluir este registro financeiro?')) {
        try {
            const response = await fetch(`${API_URL}/api/v1/financial/${id}`, {
                method: 'DELETE',
                headers: getAuthHeaders()
            });

            if (response.ok) {
                alert('Registro excluído com sucesso!');
                loadFinancialSummary();
                loadFinancialRecords();
            } else {
                alert('Erro ao excluir registro');
            }
        } catch (error) {
            console.error('Error deleting financial record:', error);
            alert('Erro de conexão');
        }
    }
}

// Apply Filters
function applyFinancialFilters() {
    const filters = {
        transaction_type: document.getElementById('filter-transaction-type').value,
        payment_status: document.getElementById('filter-payment-status').value,
        start_date: document.getElementById('filter-start-date').value,
        end_date: document.getElementById('filter-end-date').value
    };

    loadFinancialRecords(filters);
}

// Clear Filters
function clearFinancialFilters() {
    document.getElementById('filter-transaction-type').value = '';
    document.getElementById('filter-payment-status').value = '';
    document.getElementById('filter-start-date').value = '';
    document.getElementById('filter-end-date').value = '';

    loadFinancialRecords();
}
